File: lms_fe/app/_services/loan/loan.ts
import axios from "@/app/_axios/axios";
import { AmortizationPeriodType, InterestRateType, NatureOfAssets, RiskProfile } from "@/app/models/enum";
import { Criteria, LoanData, LoanDataDefaultValue } from "@/app/models/loan";

interface newLoanReq {
    name: string;
    description: string;
    criteria: loanCriteria[];
    characteristics: loanCharacteristics;
}

interface updateLoanReq {
    lid: string,
    name: string;
    description: string;
    criteria: loanCriteria[];
    characteristics: loanCharacteristics;
}


interface loanCriteria {
    risk_profile: string;
    nature_ofAssets: string;
    minimum_value_of_assets: number; // Allows decimals like 50000.75
    credit_score: number;
}

interface loanCharacteristics {
    type: string;
    amortization_period: amortizationPeriod;
    interest_rate: interestRate;
    max_value: number;
}

interface amortizationPeriod {
    duration: number;
    type: string;
}

interface interestRate {
    type: string;
    value: number;
}

interface newLoanResp{
    code: number;
    message: string;
    data: {
        lid: string;
    };
}

interface getLoanDashboardDetailResp {
    code: string,
    message: string,
    data: {
        loans:[
            {
                lid: string,
                name: string
                risk_profile: string
                interest_rate_type: string
                interest_rate_value: number 
                maximum_loan_amount: number 
            },
        ],
        total: number,
    }
}

    interface GetLoanDashboardDetailResp {
        total: number
        data: LoanData[]
    }

    // Top-level response interface
    interface loanDataResp {
        code: string; // Response code
        message: string; // Response message
        data: loanData; // Array of loan data
    }

    // Loan data interface
    interface loanData {
        lid: string; // Loan ID
        name: string; // Loan name
        description: string; // Loan description
        criteria: criteria[]; // Array of criteria
        characteristics: characteristics; // Loan characteristics
    }

    // Criteria interface
    interface criteria {
        risk_profile: string; // Risk profile (e.g., "High", "Medium")
        nature_ofAssets: string; // Nature of assets (e.g., "Secured", "Unsecured")
        minimum_value_of_assets: number; // Minimum value of assets
        credit_score: number; // Credit score
    }

    // Characteristics interface
    interface characteristics {
        type: string; // Type of loan (e.g., "Secured", "Unsecured")
        amortization_period: amortizationPeriod; // Amortization period details
        interest_rate: interestRate; // Interest rate details
        max_value: number; // Maximum loan value
    }

    // Amortization period interface
    interface amortizationPeriod {
        duration: number; // Duration of the amortization period
        type: string; // Type of amortization period (e.g., "Years", "Months")
    }

    // Interest rate interface
    interface interestRate {
        type: string; // Interest rate type (e.g., "Fixed", "Variable")
        value: number; // Interest rate value
    }


export const NewLoan = async (req: LoanData):Promise<newLoanResp> => {

    const cris = req.criteria.map((c)=>{
        return {
            risk_profile: c.riskProfile,
            nature_ofAssets: c.natureOfAssets,
            minimum_value_of_assets: c.minimumValueOfAssets,
            credit_score: c.creditScore
        }
    })

    const payload:newLoanReq = {
        name: req.name,
        description: req.description,
        criteria: cris,
        characteristics: {
            type: req.characteristics.securedOrUnsecured,
            amortization_period: {
                duration: req.characteristics.amortizationPeriod.duration,
                type: req.characteristics.amortizationPeriod.type
            },
            interest_rate: {
                type: req.characteristics.interestRate.type,
                value: Number(req.characteristics.interestRate.value)
            },
            max_value: Number(req.characteristics.maxValue)
        }
    }

    try {
        const resp = await axios.post<newLoanResp>("/loan/", payload );
        if (resp.status === 200) {
            return Promise.resolve(resp.data);
        } else {
            return Promise.reject({});
        }
    } catch (e) {
        console.error(e)
        return Promise.reject({});
    }
}

export const UpdateLoan = async (req: LoanData):Promise<newLoanResp> => {

    const cris = req.criteria.map((c)=>{
        return {
            risk_profile: c.riskProfile,
            nature_ofAssets: c.natureOfAssets,
            minimum_value_of_assets: c.minimumValueOfAssets,
            credit_score: c.creditScore
        }
    })

    const payload:updateLoanReq = {
        lid: req.id,
        name: req.name,
        description: req.description,
        criteria: cris,
        characteristics: {
            type: req.characteristics.securedOrUnsecured,
            amortization_period: {
                duration: req.characteristics.amortizationPeriod.duration,
                type: req.characteristics.amortizationPeriod.type
            },
            interest_rate: {
                type: req.characteristics.interestRate.type,
                value: Number(req.characteristics.interestRate.value)
            },
            max_value: req.characteristics.maxValue
        }
    }

    try {
        const resp = await axios.put<newLoanResp>("/loan/", payload );
        if (resp.status === 200) {
            return Promise.resolve(resp.data);
        } else {
            return Promise.reject({});
        }
    } catch (e) {
        console.error(e)
        return Promise.reject({});
    }
}

export const GetLoan = async(cid: string): Promise<LoanData> => {
    try {

        const resp = await axios.get<loanDataResp>(`/loan/${cid}`)
        if (resp.status === 200) {
            const l =  resp.data.data;
            console.log(l)

            const cris:Criteria[] = l.criteria.map((cri)=>{
                return {
                    riskProfile: RiskProfile[cri.risk_profile as keyof typeof RiskProfile],
                    natureOfAssets: NatureOfAssets[cri.nature_ofAssets as keyof typeof NatureOfAssets],
                    minimumValueOfAssets: cri.minimum_value_of_assets,
                    creditScore: cri.credit_score
                }
            })

            const loan:LoanData = {
                id: l.lid,
                name: l.name,
                description: l.description,
                criteria: cris,
                characteristics: {
                    securedOrUnsecured: l.characteristics.type,
                    amortizationPeriod: {
                        duration: l.characteristics.amortization_period.duration,
                        type: AmortizationPeriodType[l.characteristics.amortization_period.type as keyof typeof AmortizationPeriodType]
                    },
                    interestRate: {
                        type: InterestRateType[l.characteristics.interest_rate.type as keyof typeof InterestRateType],
                        value: l.characteristics.interest_rate.value
                    },
                    maxValue: l.characteristics.max_value
                }
            }

            return Promise.resolve(loan)
        } else {
            return Promise.reject({})
        }

    } catch(e){
        console.error(e)
        return Promise.reject({})
    }

}

export const GetLoanDashboardDetail = async (limit: number, offset: number): Promise<GetLoanDashboardDetailResp> => {
    try {
        
        const resp = await axios.post<getLoanDashboardDetailResp>("/loan/dashboard", {
            limit: limit, offset: offset
        });

        if (resp.status === 200) {
            const loans:LoanData[] = resp.data.data.loans.map((l)=>{
                console.log(l)
                return {
                    ...LoanDataDefaultValue,
                    id: l.lid,
                    name: l.name,
                    description: "",
                    characteristics: {
                        ...LoanDataDefaultValue.characteristics,
                        interestRate: {
                            type: InterestRateType[l.interest_rate_type as keyof typeof InterestRateType],
                            value: l.interest_rate_value
                        },
                        maxValue: l.maximum_loan_amount
                    },
                    criteria: [
                        {
                            riskProfile: RiskProfile[l.risk_profile as keyof typeof RiskProfile],
                            natureOfAssets: NatureOfAssets.UNSECURED,
                            minimumValueOfAssets: 0,
                            creditScore: 0, 
                        },
                    ],
                }
            })

            return Promise.resolve({
                total: resp.data.data.total,
                data: loans,
            })

        } else {
            return Promise.reject({})
        }

    } catch (e) {
        console.error(e)
        return Promise.reject({});
    }
}

export const DeleteLoan = async (lid: string): Promise<string> => {
    try {
        const resp = await axios.delete(`/loan/${lid}`);
        if (resp.status === 200) {
            return Promise.resolve("success");
        } else {
            return Promise.reject("");
        }
    } catch (e) {
        console.error(e)
        return Promise.reject("");
    }
}

File: lms_be/internal/models/loan.go
package models

import (
	"gorm.io/gorm/schema"
)

type LoanData struct {
	Lid             string          `json:"lid"`
	Name            string          `json:"name"`
	Description     string          `json:"description"`
	Criteria        []Criteria      `json:"criteria"`
	Characteristics Characteristics `json:"characteristics"`
}

type LoanRepo struct {
	Lid                    string  `gorm:"lid"`
	Name                   string  `gorm:"name"`
	Description            string  `gorm:"description"`
	Criteria               string  `gorm:"criteria"`
	CharacteristicType     string  `gorm:"characteristic_type"`
	AmortizationDuration   int64   `gorm:"amortization_duration"`
	AmortizationPeriodType string  `gorm:"amortization_period_type"`
	InterestRateType       string  `gorm:"interest_rate_type"`
	InterestRateValue      float64 `gorm:"interest_rate_value"`
	MaximumLoanAmount      int64   `gorm:"maximum_loan_amount"`
}

func (l LoanRepo) TableName(namer schema.Namer) string {
	tbName := "loans"
	if namer == nil {
		return tbName
	}
	return namer.TableName(tbName)
}

type Criteria struct {
	RiskProfile          string  `json:"risk_profile"`
	NatureOfAssets       string  `json:"nature_ofAssets"`
	MinimumValueOfAssets float64 `json:"minimum_value_of_assets"`
	CreditScore          int     `json:"credit_score"`
}

type CriteriaRepo struct {
	Cid           string `gorm:"cid"`
	RiskProfile   string `gorm:"risk_profile"`
	AssetNature   string `gorm:"asset_nature"`
	MinAssetValue int64  `gorm:"min_asset_value"`
	CreditScore   int64  `gorm:"credit_score"`
}

func (c CriteriaRepo) TableName(namer schema.Namer) string {
	tbName := "loan_criteria"
	if namer == nil {
		return tbName
	}
	return namer.TableName(tbName)
}

type Characteristics struct {
	Type               string             `json:"type"` // SecuredOrUnsecured
	AmortizationPeriod AmortizationPeriod `json:"amortization_period"`
	InterestRate       InterestRate       `json:"interest_rate"`
	MaxValue           int64              `json:"max_value"`
}

type AmortizationPeriod struct {
	Duration int64  `json:"duration"`
	Type     string `json:"type"`
}

type InterestRate struct {
	Type  string  `json:"type"`
	Value float64 `json:"value"`
}

type NewLoanReq struct {
	Name            string          `json:"name"`
	Description     string          `json:"description"`
	Criteria        []Criteria      `json:"criteria"`
	Characteristics Characteristics `json:"characteristics"`
}

type NewLoanResp struct {
	Lid string `json:"lid"`
}

type UpdateLoanReq struct {
	Lid string `json:"lid"`
	NewLoanReq
}

type UpdateLoanResp struct {
	Lid string `json:"lid"`
}

type GetLoanDashboardDetailReq struct {
	Limit  uint64 `json:"limit"`
	Offset uint64 `json:"offset"`
}

type GetLoanDashboardDetailData struct {
	Lid               string `json:"lid"`
	Name              string `json:"name"`
	RiskProfile       string `json:"risk_profile"`
	InterestRateType  string `json:"interest_rate_type"`
	InterestRateValue string `json:"interest_rate_value"`
	MaximumLoanAmount string `json:"maximum_loan_amount"`
}

type GetLoanDashboardDetailResp struct {
	Loans []GetLoanDashboardDetailData `json:"loans"`
	Total int64                        `json:"total"`
}

type GetLoanDashboardDetailRepo struct {
	Lid               string `gorm:"lid"`
	Name              string `gorm:"name"`
	Criteria          string `gorm:"criteria"`
	InterestRateType  string `gorm:"interest_rate_type"`
	InterestRateValue string `gorm:"interest_rate_value"`
	MaximumLoanAmount string `gorm:"maximum_loan_amount"`
}

func (l GetLoanDashboardDetailRepo) TableName(namer schema.Namer) string {
	tbName := "loans"
	if namer == nil {
		return tbName
	}
	return namer.TableName(tbName)
}
